File: 871-keys-and-rooms/keys-and-rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        vector<bool> visited(n, false);
        queue<int> q;

        q.push(0);
        visited[0] = true;

        while(!q.empty()) {
            int node = q.front();
            q.pop();

            for(int neighbours : rooms[node]) {
                if(!visited[neighbours]) {
                    visited[neighbours] = true;
                    q.push(neighbours);
                }
            }
        }
        for(bool isVisited : visited) {
            if(!isVisited) 
                return false;
        }
        return true;
    }
};

File: 871-keys-and-rooms/keys-and-rooms.java
class Solution {
    public void DFS(List<List<Integer>> grid, boolean[] visited, int node) {
        visited[node] = true;

        for (int neighbour : grid.get(node)) {
            if(visited[neighbour] == false) {
                DFS(grid, visited, neighbour);
            }
        }
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int n = rooms.size();
        boolean[] visited = new boolean[n];

        DFS(rooms, visited, 0);
        for(boolean node : visited) {
            if(node == false)
                return false;
        }
        return true;
    }
}